### PR TITLE
*: Support the MySQL 8.0 #176

### LINKED ISSUE #176: [Feature] support MySQL 8.0
**Is your feature request related to a problem? Please describe.**

**Describe the solution you'd like**
Support MySQL 8.0 

**Describe alternatives you've considered**

**Additional context**

Mysql 8.0 create user do not like Mysql 5.7 , and so on

### PR DESCRIPTION
**Pull request by acekingke**
### What type of PR is this?

/enhancement
### Which issue(s) this PR fixes?

Fixes #176 

### What this PR does?

Summary:

### CODE REVIEW

File: utils/constants.go [hunk 1]
```diff
@@ -23,11 +23,13 @@ var (
 	// MySQLTagsToSemVer maps simple version to semver versions
 	MySQLTagsToSemVer = map[string]string{
 		"5.7": "5.7.33",
+		"8.0": "8.0",
```
Comment (hustjieke) [inline utils/constants.go:26]: `8.0`-->`8.0.25`?
or we change `5.7.33` to `5.7` ?

File: utils/constants.go [hunk 2]
```diff
@@ -23,11 +23,13 @@ var (
...
 	}
 
 	// MysqlImageVersions is a map of supported mysql version and their image
 	MysqlImageVersions = map[string]string{
 		"5.7.33": "percona/percona-server:5.7.33",
+		"8.0":    "percona/percona-server:8.0.25",
```
Comment (hustjieke) [inline utils/constants.go:32]: `8.0`-->`8.0.25`
or we change `5.7.33` to `5.7` ?

**Comment by andyli029** [comment]
Notice:  xtrabackup 8.0 is different to xtrabackup 2.4

**Comment by acekingke** [comment]
> Notice: xtrabackup 8.0 is different to xtrabackup 2.4

Use the following build image：
`docker build --build-arg XTRABACKUP_PKG=percona-xtrabackup-80  -f  Dockerfile.sidecar -t ${SIDECAR_IMG} .`

@andyli029 @runkecheng

**Comment by runkecheng** [comment]
```
 mysql57_conf = """
            "mysql": {
                    "version": "mysql57",
                    "admin": "root",
                    "passwd": "",
                    "host": "localhost",
                    "port": %s,
                    "basedir": "%s",
                    "defaults-file": "%s",
                    "master-sysvars": "tokudb_fsync_log_period=default;sync_binlog=default;innodb_flush_log_at_trx_commit=default",
                    "slave-sysvars":  "tokudb_fsync_log_period=1000;sync_binlog=1000;innodb_flush_log_at_trx_commit=1",
                    "ping-timeout": %s,
                    "admit-defeat-ping-count": 3,
                    "semi-sync-timeout-for-two-nodes": %s
            },
              ......

        mysql80_conf = """
            "mysql": {
                    "version": "mysql80",
                    "admin": "root",
                    "passwd": "",
                    "host": "localhost",
                    "port": %s,
                    "basedir": "%s",
                    "defaults-file": "%s",
                    "master-sysvars": "sync_binlog=default;innodb_flush_log_at_trx_commit=default",
                    "slave-sysvars":  "sync_binlog=1000;innodb_flush_log_at_trx_commit=1",
                    "ping-timeout": %s,
                    "admit-defeat-ping-count": 3,
                    "semi-sync-timeout-for-two-nodes": %s
            },
           ......
```

**Comment by acekingke** [comment]
> Remove parameters of tokudb
> 
> ```
>  mysql57_conf = """
>             "mysql": {
>                     "version": "mysql57",
>                     "admin": "root",
>                     "passwd": "",
>                     "host": "localhost",
>                     "port": %s,
>                     "basedir": "%s",
>                     "defaults-file": "%s",
>                     "master-sysvars": "tokudb_fsync_log_period=default;sync_binlog=default;innodb_flush_log_at_trx_commit=default",
>                     "slave-sysvars":  "tokudb_fsync_log_period=1000;sync_binlog=1000;innodb_flush_log_at_trx_commit=1",
>                     "ping-timeout": %s,
>                     "admit-defeat-ping-count": 3,
>                     "semi-sync-timeout-for-two-nodes": %s
>             },
>               ......
> 
>         mysql80_conf = """
>             "mysql": {
>                     "version": "mysql80",
>                     "admin": "root",
>                     "passwd": "",
>                     "host": "localhost",
>                     "port": %s,
>                     "basedir": "%s",
>                     "defaults-file": "%s",
>                     "master-sysvars": "sync_binlog=default;innodb_flush_log_at_trx_commit=default",
>                     "slave-sysvars":  "sync_binlog=1000;innodb_flush_log_at_trx_commit=1",
>                     "ping-timeout": %s,
>                     "admit-defeat-ping-count": 3,
>                     "semi-sync-timeout-for-two-nodes": %s
>             },
>            ......
> ```
image
MySQL8 initTokoDB must be false

**Comment by acekingke** [comment]
After add data, then clone int, sacle out from 2 to 3, then got error:

```
2022-01-20T13:51:23.689045+08:00 14 [ERROR] [MY-010584] [Repl] Slave I/O for channel '': error connecting to 
plugin 'caching_sha2_password' reported error: Authentication requires secure connection. Error_code: MY-002061

2022-01-20T13:52:23.690476+08:00 14 [ERROR] [MY-010584] [Repl] Slave I/O for channel '': error connecting to master 'radondb_repl@sample-mysql-1.sample-mysql.default:3306' - retry-time: 60 retries: 2 message: Authentication plugin 'caching_sha2_password' reported error: Authentication requires secure connection. Error_code: MY-002061
```

**Comment by andyli029** [comment]
Reference:
https://www.percona.com/doc/percona-server/8.0/release-notes/Percona-Server-8.0.25-15.html
https://dev.mysql.com/doc/relnotes/mysql/8.0/en/news-8-0-25.html

**Review by runkecheng — COMMENTED** [review]
cluster,config,sidecar,utils -> *

**Review by andyli029 — COMMENTED** [review]
Check the version about the Minor of 8.0.

**Comment by runkecheng** [comment]
Add a tutorial and sample yaml for mysql 8.

File: mysqlcluster/syncer/variables.go [hunk 3]
```diff
@@ -17,86 +17,92 @@ limitations under the License.
...
+	"relay_log":                 "/var/lib/mysql/mysql-relay-bin",
+	"relay_log_index":           "/var/lib/mysql/mysql-relay-bin.index",
+	"master_info_repository":    "TABLE",
+	"relay_log_info_repository": "TABLE",
+	"slow_query_log":            "1",
+	"tmp_table_size":            "32M",
+	"tmpdir":                    "/var/lib/mysql",
+}
+
+var mysqlSpecialConfig = map[string]string{
```
Comment (runkecheng) [inline mysqlcluster/syncer/variables.go:50]: mysqlSpecialConfig -> mysql8Configs is better?
[ref](https://github.com/bitpoke/mysql-operator/blob/5a9fd7692c5deb8e264f65df57738c2afc9e694c/pkg/controller/mysqlcluster/internal/syncer/config_map.go#L102)

File: utils/constants.go [hunk 4]
```diff
@@ -28,12 +28,14 @@ var (
 	// MySQLTagsToSemVer maps simple version to semver versions
 	MySQLTagsToSemVer = map[string]string{
 		"5.7": "5.7.34",
+		"8.0": "8.0.25",
```
Comment (runkecheng) [inline utils/constants.go:31]: why 8.0.25?

Comment (acekingke) [reply utils/constants.go:31]: percona docker image version, just start from 8.0.25

File: config/samples/mysql_v1alpha1_mysqlcluster_mysql8.yaml [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+    resources:
+      requests:
+        cpu: 10m
+        memory: 32Mi
+      limits:
+        cpu: 100m
+        memory: 128Mi
+  podPolicy:
+    imagePullPolicy: Always
+    sidecarImage: # fill here with image sidecar for mysql8
```
Comment (runkecheng) [inline config/samples/mysql_v1alpha1_mysqlcluster_mysql8.yaml:55]: radondb/mysql80-sidecar:v2.1.4

Comment (runkecheng) [inline config/samples/mysql_v1alpha1_mysqlcluster_mysql8.yaml:55]: radondb/mysql80-sidecar:v2.1.4

Comment (runkecheng) [reply config/samples/mysql_v1alpha1_mysqlcluster_mysql8.yaml:55]: Need push manually.

**Comment by GuishengZhou** [comment]
[mysqld]
default_authentication_plugin=mysql_native_password
port = 3306
basedir = /opt/mysql
datadir =/data/mysql/
socket = /tmp/mysql.sock
log-error =error.txt

server_id=50165
binlog_format = row
log-bin=mysql-bin
open_files_limit=65536
max_connections=1024
core-file

##semi-sync
#plugin-load="semisync_master.so;semisync_slave.so"
#rpl_semi_sync_master_enabled=OFF
#rpl_semi_sync_slave_enabled=ON
#rpl_semi_sync_master_wait_no_slave=ON
#rpl_semi_sync_master_timeout=1000000000000000000


use  file  to init default-file 
##repl
skip-slave-start
gtid-mode = ON
enforce-gtid-consistency = ON
slave_parallel_type = LOGICAL_CLOCK
slave_parallel_workers=128
log-slave-updates

#innodb
innodb_buffer_pool_size=4G

sql_mode=NO_ENGINE_SUBSTITUTION,STRICT_TRANS_TABLES
slave_rows_search_algorithms=INDEX_SCAN,HASH_SCAN
lower_case_table_names=1

File: mysqlcluster/syncer/variables.go (hunk 3 above)
Comment (acekingke) [reply mysqlcluster/syncer/variables.go]: this configs use both by mysql5.7 and mysql8, but mysql8 should deal it in special way.

File: config/samples/mysql_v1alpha1_mysqlcluster_mysql8.yaml (hunk 5 above)
Comment (acekingke) [reply config/samples/mysql_v1alpha1_mysqlcluster_mysql8.yaml:55]: need create the sidecar image.

**Comment by andyli029** [comment]
link：v2.2.0

**Pull request closed without merging**